packages/powersync_core/lib/src/attachments/attachment_queue_service.dart

```diff
@@ -0,0 +1,443 @@
+  }) async {
+    final resolvedId = id ?? await generateAttachmentId();
+
+    final filename = await resolveNewAttachmentFilename(
+      resolvedId,
+      fileExtension,
+    );
+
+    // Write the file to the filesystem.
+    final fileSize = await _localStorage.saveFile(filename, data);
+
+    return await _attachmentsService.withContext((attachmentContext) async {
+      return await _db.writeTransaction((tx) async {
+        final attachment = Attachment(
+          id: resolvedId,
+          filename: filename,
+          size: fileSize,
+          mediaType: mediaType,
+          state: AttachmentState.queuedUpload,
+          localUri: filename,
+          metaData: metaData,
+        );
+
+        // Allow consumers to set relationships to this attachment ID.
+        await updateHook(tx, attachment);
+
+        return await attachmentContext.upsertAttachment(attachment, tx);
+      });
+    });
+  }
+
+  /// Queues an attachment for delete.
+  /// The default implementation assumes the attachment record already exists locally.
+  Future<Attachment> deleteFile({
+    required String attachmentId,
+    required Future<void> Function(
+            SqliteWriteContext context, Attachment attachment)
+        updateHook,
+  }) async {
+    return await _attachmentsService.withContext((attachmentContext) async {
+      final attachment = await attachmentContext.getAttachment(attachmentId);
+      if (attachment == null) {
+        throw Exception(
+          'Attachment record with id $attachmentId was not found.',
+        );
+      }
+
+      return await _db.writeTransaction((tx) async {
+        await updateHook(tx, attachment);
+        return await attachmentContext.upsertAttachment(
+          attachment.copyWith(
+            state: AttachmentState.queuedDelete,
+            hasSynced: false,
+          ),
+          tx,
+        );
+      });
+    });
+  }
+
+  /// Removes all archived items.
+  Future<void> expireCache() async {
+    await _attachmentsService.withContext((context) async {
+      bool done;
+      do {
+        done = await _syncingService.deleteArchivedAttachments(context);
+      } while (!done);
+    });
+  }
+
+  /// Clears the attachment queue and deletes all attachment files.
+  Future<void> clearQueue() async {
+    await _attachmentsService.withContext((context) async {
+      await context.clearQueue();
+    });
+    await _localStorage.clear();
+  }
+
+  /// Cleans up stale attachments.
+  Future<void> _verifyAttachments(AttachmentContext context) async {
+    final attachments = await context.getActiveAttachments();
+    final List<Attachment> updates = [];
+
+    for (final attachment in attachments) {
+      // Only check attachments that should have local files
+      if (attachment.localUri == null) {
+        // Skip attachments that don't have localUri (like queued downloads)
+        continue;
+      }
+
+      final exists = await _localStorage.fileExists(attachment.localUri!);
```

Comment:
It's only an iOS related issue, so slightly less relevant here, but since we are working on this package now. It could be good to implement the changes from here in this package https://github.com/powersync-ja/powersync-swift/pull/74.